pkg/master/controller.go

```diff
@@ -323,12 +396,30 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
 			}
 		}
 	}
-	for _, addr := range sub.Addresses {
-		if addr.IP == ip {
-			ipCorrect = len(sub.Addresses) <= count
-			break
+
+	ipCorrect = true
+	if len(sub.Addresses) != len(expectedIPs) {
+		ipCorrect = false
+	} else {
+		// check the actual content of the addresses
+		// present addrs is used as a set (the keys) and to indicate if a
+		// value was already found (the values)
+		presentAddrs := make(map[string]bool, len(expectedIPs))
+		for _, ip := range expectedIPs {
+			presentAddrs[ip] = false
+		}
+
+		// currentAddrs == expectedAddr iff (len(ca) == len(ea) && ca <= ea && ca has no duplicates)
```

Comment:
confusing comment. We actually don't check the uniqueness, but assume it as a pre-condition.